src/NServiceBus.Core/InitializableEndpoint.cs

```diff
@@ -47,35 +47,58 @@ public InitializableEndpoint(SettingsHolder settings, IContainer container, List
             var transportInfrastructure = transportDefinition.Initialize(settings, connectionString);
             settings.Set<TransportInfrastructure>(transportInfrastructure);
 
+            var isSendOnlyEndpoint = settings.Get<bool>("Endpoint.SendOnly");
+            var receiving = new ReceiveComponent(settings.EndpointName(), isSendOnlyEndpoint, transportInfrastructure);
+
+
+            var receiveConfiguration = receiving.Configure(settings);
+
+            //note: remove once settings.LogicalAddress() , .LocalAddress() and .InstanceSpecificQueue() has been obsoleted
+            settings.Set<ReceiveConfiguration>(receiveConfiguration);
+
             // use GetOrCreate to use of instances already created during EndpointConfiguration.
             var routing = new RoutingComponent(
                 settings.GetOrCreate<UnicastRoutingTable>(),
                 settings.GetOrCreate<DistributionPolicy>(),
                 settings.GetOrCreate<EndpointInstances>(),
                 settings.GetOrCreate<Publishers>());
-            routing.Initialize(settings, transportInfrastructure, pipelineSettings);
+            routing.Initialize(settings, transportInfrastructure, pipelineSettings, receiveConfiguration);
 
-            var featureStats = featureActivator.SetupFeatures(container, pipelineSettings, routing);
+            var featureStats = featureActivator.SetupFeatures(container, pipelineSettings, routing, receiveConfiguration);
 
             pipelineConfiguration.RegisterBehaviorsInContainer(container);
 
             DisplayDiagnosticsForFeatures.Run(featureStats);
 
             container.ConfigureComponent(b => settings.Get<Notifications>(), DependencyLifecycle.SingleInstance);
 
+            var eventAggregator = new EventAggregator(settings.Get<NotificationSubscriptions>());
+            var mainPipeline = new Pipeline<ITransportReceiveContext>(builder, pipelineConfiguration.Modifications);
+            var pipelineCache = new PipelineCache(builder, settings);
+            var mainPipelineExecutor = new MainPipelineExecutor(builder, eventAggregator, pipelineCache, mainPipeline);
+            var errorQueue = settings.ErrorQueueAddress();
+
+            var receiveRuntime = await receiving.InitializeRuntime(receiveConfiguration,
```

Comment:
If we could replace dependency on `MainPipelineExecutor` with `Func<IBuilder, MessageContext, Task> onMessage` then we would be able, in theory, to run the `ReceiveRuntime` without the whole endpoint & pipeline.

The only missing thing to replace the NSB.Raw would be an abstraction for recoverability: `RecoverabilityExecutor` and its factory are not replaceable.

Reply:
I changed to the existing `IPipelineExecutor` which should get the job done given that its  `Task OnMessage(MessageContext c)`